Simulator with large range inputs

ckyiricky opened the pull request:
This branch is derived from branch simulator-tests, so compared to master, it could have some overlapped changes with the tests PR. 
[Add] Add range input generator
    Add range input generator for larger scale tests
    Add shell script to run all algorithms
    Update one algorithm name in algorithms.go to be consistent with others
[Update] Update process to multi-threaded
    Update process to multi-threaded which can process large data input files
    Update local-shared algorithm that only zones with local SG > 1 endpoint
    can be candidates(give endpoints to others)
    Update local-shared algorithm that urgent zones will no longer be added
    to 'needed list' separately
[Add] Add result parser 
[Update] Optimize local-shared algorithm with corner cases.
[Update] Evaluation mechanism with only positive deviation

Commit:
[Add] Add range input generator
Add range input generator for larger scale tests
Add shell script to run all algorithms
Update one algorithm name in algorithms.go to be consistent with others

Commit:
[Add/Update] Update process to multi-threaded
Add result parser python script
Update process to multi-threaded which can process large data input files
Update local-shared algorithm that only zones with local SG > 1 endpoint
can be candidates(give endpoints to others)
Update local-shared algorithm that urgent zones will no longer be added
to 'needed list' separately

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -225,3 +228,9 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
+// detect whether a zone is valid to contribute endpoints to other zones
+func validContributor(deviation float64, zoneName string, sliceGroup types.EndpointSliceGroup) bool {
+	// sliceGroup.Composition[zoneName] is guaranteed to be non-nil by caller
+	return deviation > 0 && sliceGroup.Composition[zoneName].Number > 1
```
I'd feel better if there were a nil check on `sliceGroup.Composition[zoneName]` here since this can be called from multiple places.

ckyiricky replied:
added nil check

robscott commented on process/input-parser.go:
```diff
@@ -28,13 +28,14 @@ import (
-// parseInput parses input csv file to instances of inputData and returns a
-// slice of them
-func parseInput(file string) ([]inputData, error) {
+// paserInput parses input csv file to instances of inputData and puts them into
```
```suggestion
// parseInput parses an input csv file to instances of inputData and puts them into
```

ckyiricky replied:
changed

robscott commented on process/input-parser.go:
```diff
@@ -47,21 +48,33 @@ func parseInput(file string) ([]inputData, error) {
+			data, done, err = readOneRow(zoneNames, reader)
+			if done {
+				break
+			}
```
Would something like this work?
```suggestion
		for data, done, err := readOneRow(zoneNames, reader); !done {
```

ckyiricky replied:
changed to an embedded way: 
`for data, done, rerr := readOneRow(zoneNames, reader); !done; data, done, rerr = readOneRow(zoneNames, reader) {`
Need to call it again in the 3rd section and use another name other than err to avoid shadow previous err variable.

robscott commented on process/output-parser.go:
```diff
@@ -27,13 +27,16 @@ import (
+			klog.Errorf("close output file %s with an error %v", file, cerr)
+		}
 		if err == nil {
 			err = cerr
```
maybe this should just be `return cerr`

robscott commented on process/output-parser.go:
```diff
@@ -48,7 +51,11 @@ func parseResult(file string, outputArray []outputData) (err error) {
+		rowData, more := <-outputQueue
+		if !more {
+			break
+		}
```
Would this work?
```suggestion
	for rowData, more := <-outputQueue; !more {
```

ckyiricky replied:
similar to above, changed to an embedded way.

robscott commented on process/process.go:
```diff
@@ -66,33 +71,49 @@ type outputData struct {
+	defer close(errC)
+
+	go func() {
+		defer close(outputQueue)
```
Why is this being closed inside the goroutine while the error channel is closed outside of it?

ckyiricky replied:
changed

robscott commented on process/process.go:
```diff
@@ -66,33 +71,49 @@ type outputData struct {
+			rowData, more := <-inputQueue
+			if !more {
+				break
+			}
```
```suggestion
		for rowData, more := <-inputQueue; !more {
```

ckyiricky replied:
changed as the other two places.

robscott commented on process/process.go:
```diff
@@ -66,33 +71,49 @@ type outputData struct {
+			outputQueue <- outputData{name: rowData.name,
+				endpoints:      model.GetNumberOfEndpoints(),
+				endpointSlices: model.GetNumberOfEndpointSlices(),
+				result:         simRes}
```
```suggestion
			oData, err := runSimulation(model, rowData.Zones)
			if err == nil {
				outputQueue <- oData
			}
```

ckyiricky replied:
done

robscott commented on process/range-input-generator.py:
```diff
@@ -0,0 +1,66 @@
+"""
```
I think this should be a separate dir away from the go code.

robscott replied:
Also should add a note somewhere that this requires python 3.

robscott commented on run-all.sh:
```diff
@@ -0,0 +1,20 @@
+# Copyright 2020 Google LLC
```
Since both this and the python script are a new way to interact with this project, this PR should also include some docs updates.

robscott replied:
This file could use `chmod +x`

robscott replied:
I think this file should also start with 

```
#!/bin/bash
```

ckyiricky replied:
TODO: doc update in future PRs

robscott reviewed (commented):
Thanks for the work on this! This mostly LGTM, still need to dig into the python script though.

robscott commented on run-all.sh:
```diff
@@ -0,0 +1,20 @@
+
+for alg in ${algs[@]}; do
+    echo "Running $alg"
+    go run main.go -input=example/range-input.csv -alg=$alg -output=example/$alg-range-output.csv
```
`open example/range-input.csv: no such file or directory`

robscott replied:
Probably want to add a check with bash for the existence of this file before starting this loop. If it doesn't exist you could exit and log that the python generation script needs to run first.

robscott replied:
Also as a sidebar, would probably be good to use a directory other than `example` here and/or add `range-input.csv` to `.gitignore`

ckyiricky replied:
added

robscott commented on process/range-input-generator.py:
```diff
@@ -0,0 +1,66 @@
+from itertools import combinations 
+
+# Print iterations progress
+def printProgressBar (iteration, total, progress="Progress", length = 100, fill = '█', printEnd = "\r"):
```
As much as I personally prefer camel case, snake case is the standard for Python. https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

ckyiricky replied:
changed

robscott commented on process/range-input-generator.py:
```diff
@@ -0,0 +1,66 @@
+    nodes_comb = list(combinations(range(1, 11), 3))
+    for index, nodes in enumerate(nodes_comb):
+        generateOneSection(writer, nodes, suffix='-'+str(nodes), bar=False)
+        printProgressBar(index+1, len(nodes_comb), progress="nodes iteration")
```
This ends up feeling a bit inconsistent with output for the other section. Seems like the names should be more similar (0-100 endpoints, 100-1000 endpoints) and ideally the progress bars would line up more closely. Overall this is a really cool interface though.

```
nodes iteration: |████████████████████████████████████████████████████████████████████████████████████████████████████| 100.00% Complete
100-1000 endpoints [30, 30, 30] nodes, step: 7: |██████████████████████████████████████████████████████████████████████| 100.00% Complete
```

ckyiricky replied:
changed

robscott commented on process/range-input-generator.py:
```diff
@@ -0,0 +1,66 @@
+    # generate range input: 3 zones
+    # nodes = 30 for every zone
+    # endpoints range from 100-1000 with step = 7
+    generateOneSection(writer, ep_range=[100, 1001], step = 7, suffix='-high')
```
I just looked at the first and last 100 lines of this output and it seems like the deduping might be a little too eager. As an example, here's the last lines of my output:

```
349494-high,30 968,30 975,30 982
349495-high,30 968,30 975,30 989
349496-high,30 968,30 975,30 996
349497-high,30 968,30 982,30 989
349498-high,30 968,30 982,30 996
349499-high,30 968,30 989,30 996
349500-high,30 975,30 982,30 989
349501-high,30 975,30 982,30 996
349502-high,30 975,30 989,30 996
349503-high,30 982,30 989,30 996
```

Maybe this is covered by another test case elsewhere, but it seems like it would also be useful to test examples where some zones had the same number of endpoints like 996, 996, 989.

ckyiricky replied:
fixed, need further discussion though

robscott reviewed (commented):
Thanks for all the work on this! Got a chance to run the python and shell scripts and both work well on my computer as well.

ckyiricky commented:
Add a result parser for large output
Change some coding style

Commit:
[Update] Optimize local-shared algorithm with corner cases
Update local-shared algorithm to maintain max deviation less than a
threshold
Update local-shared test with more corner cases
Update evalution mechanism to ignore negative deviation(less traffic
than expected)
Update readme for large dataset input usage
Update move scripts to hack dir

ckyiricky commented:
[Update] Optimize local-shared algorithm with corner cases
    Update local-shared algorithm to maintain max deviation less than a
    threshold
    Update local-shared test with more corner cases
    Update evaluation mechanism to ignore negative deviation(less traffic
    than expected)
    Update readme for large dataset input usage

robscott commented on README.md:
```diff
@@ -12,13 +12,33 @@ Kubernetes. This specific project will include:
+### Multiple algorithms usage
+`sh ./run-all.sh [input-file]`
+
+input-file is optional, ./data/range-input.csv is set as default.
```
Maybe note that this file can be generated by `range-input-generator.py` described below.

ckyiricky replied:
added

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -229,11 +245,163 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
+}
+
+// check if endpoints in receiveZone have invalid deviation
+func (alg LocalSharedSliceAlgorithm) invalidDeviation(receiveZone string, region types.RegionInfo, sliceGroups map[string]types.EndpointSliceGroup, delta int) bool {
```
Maybe something like `deviationAboveThreshold()`?

ckyiricky replied:
done

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -114,11 +118,10 @@ func (alg LocalSharedSliceAlgorithm) CreateSliceGroups(region types.RegionInfo)
 	receiverPool.SliceGroups = sliceGroups
 
-	succ, err := alg.balanceSliceGroups(&endpointsNeeded, &endpointsNeededUrgent, sliceGroups, &availablePool, &receiverPool)
+	succ, err := alg.balanceSliceGroups(&endpointsNeeded, &endpointsNeededUrgent, region, sliceGroups, &availablePool, &receiverPool)
```
Do we need both success and err here? If so we should handle err before success.

ckyiricky replied:
handle the error check at first.

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -147,7 +149,15 @@ func (alg LocalSharedSliceAlgorithm) balanceSliceGroups(endpointsNeeded *endpoin
 	// take the ceil, if the deviation > 0, we have to make sure there is at
 	// least one endpoint in this shared SG (avoid 0.x ending up with 0)
-	mergedED.deviation = int(math.Ceil(mergedDeviation))
+	if mergedDeviation >= 1 {
```
sum of expectedEndpoints?

ckyiricky replied:
changed

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -193,7 +207,7 @@ func (alg LocalSharedSliceAlgorithm) balanceSliceGroups(endpointsNeeded *endpoin
 		// candidate is guaranteed to have a local owned SG, omit the second
 		// returned value
-		deviation, _ := getEndpointsDeviation(availablePool.Region, availablePool.SliceGroups, candidate)
+		deviation, _ := getEndpointsDeviation(region, sliceGroups, candidate)
```
Are we using the second return value anywhere?

ckyiricky replied:
added usage of the second value, which should always be true, if its not, return error (unexpected)

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -229,11 +245,163 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
+	var urgentZones []string
+	for receiverPool.Len() > 0 {
+		receiveZone := receiverPool.ZoneNames[0]
+		if !alg.invalidDeviation(receiveZone, region, sliceGroups, 0) {
```
`deviationAboveThreshold()`

robscott replied:
Maybe explain why you're breaking here?

ckyiricky replied:
added

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -229,11 +245,163 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
+		heap.Pop(receiverPool)
+	}
+
+	// get extra endpoints from zones
```
Maybe add a bit more of an explanation of what extra endpoints are

ckyiricky replied:
added

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -229,11 +245,163 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
+	extraEndpointsNumber := 0
+	if len(urgentZones) > 0 {
+		// if extraEndpointsNumber >= len(urgentZones), we just assign one
+		// enpoint to each urgent zone, their deviation will go below threshold
```
```suggestion
		// endpoint to each urgent zone, their deviation will go below threshold
```

ckyiricky replied:
fixed

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -229,11 +245,163 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
+	}
+	// continuously check if the extra endpoints can help build a valid shared
+	// sliceGroup
+	for !alg.trafficDeviationAdequate(urgentZones, availablePool.Region, sliceGroups, extraEndpointsNumber) {
```
Function is asking if there are enough extra endpoints for a shared slice to get deviation below threshold. Maybe a function name like `enoughExtraEndpoints` or `sufficientExtraEndpoints` would be clearer.

ckyiricky replied:
changed to sufficientExtraEndpointsForSharedSlice

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -229,11 +245,163 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
+		extraEndpointsNumber += num
+	}
+	// continuously check if the extra endpoints can help build a valid shared
+	// sliceGroup
```
```suggestion
	// rebalance endpoints until deviations are below threshold or nothing left in availablePool
```

ckyiricky replied:
changed

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -147,7 +149,15 @@ func (alg LocalSharedSliceAlgorithm) balanceSliceGroups(endpointsNeeded *endpoin
+		mergedDeviation = math.Ceil(mergedDeviation*1000) / 1000
+		mergedED.deviation = int(math.Round(mergedDeviation))
+	} else {
+		mergedED.deviation = int(math.Ceil(mergedDeviation))
```
Ceiling for 1.0000001 probably doesn't make sense

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -147,7 +149,15 @@ func (alg LocalSharedSliceAlgorithm) balanceSliceGroups(endpointsNeeded *endpoin
+		// precision lose, i.e. 1.5 will end up with 1.49999998 that be rounded
+		// to 1 instead of 2
+		mergedDeviation = math.Ceil(mergedDeviation*1000) / 1000
+		mergedED.deviation = int(math.Round(mergedDeviation))
```
I don't think this makes sense here but it may make more sense below for ceil.

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -229,11 +245,163 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
+func (alg LocalSharedSliceAlgorithm) getMoreEndpoints(availablePool *ZonePriorityQueue, extraEndpoints map[string]int, urgentZones []string) bool {
+	sliceGroups := availablePool.SliceGroups
+	region := availablePool.Region
+	extraEndpointsNumber := 0
```
Add a comment to describe how this is used.

ckyiricky replied:
added

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -229,11 +245,163 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
+
+// when it failed to create a shared sliceGroup for urgent zones to achieve
+// deviation less than threshold, it gets more endpoints from other zones.
+func (alg LocalSharedSliceAlgorithm) getMoreEndpoints(availablePool *ZonePriorityQueue, extraEndpoints map[string]int, urgentZones []string) bool {
```
This is a hard one to name, but maybe one of the following would work:

- `getExtraEndpoints`
- `getSharedSliceEndpoints`
- `getEndpointsForSharedSlice`
- `getExtraEndpointsForSharedSlice`

ckyiricky replied:
changed to `getExtraEndpointsForSharedSlice`

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -225,3 +244,164 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
+}
+
+// when it failed to create a shared sliceGroup for urgent zones to achieve
+// deviation less than threshold, it gets more endpoints from other zones.
```
```suggestion
// [function name] attempts to get extra endpoints that could be used for a shared slice to get deviation below threshold. Returns false if it can't.
```

ckyiricky replied:
changed

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -225,3 +244,164 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
+	}
+	for zone, number := range extraEndpoints {
+		updateSGComposition(sharedSG, zone, number, 1)
+	}
```
Some comments would be helpful here to explain what `updateSGComposition` is responsible for here. Is it taking endpoints previously allocated to the zone's slice group and giving them to the shared slice group?

ckyiricky replied:
added

robscott commented on data/range-input-generator.py:
```diff
@@ -0,0 +1,69 @@
+#! /usr/bin/env python3
```
I think it would good to separate the scripts to generate data from where the data is stored. Maybe these scripts could move into a new `scripts` or `hack` director.

ckyiricky replied:
moved to hack dir

robscott reviewed (commented):
Thanks for your work on this! Some great improvements here.

ckyiricky commented:
Update function names
Move scripts to hack dir
Add more comments

robscott commented on process/input-parser.go:
```diff
@@ -28,13 +28,14 @@ import (
-// parseInput parses input csv file to instances of inputData and returns a
-// slice of them
-func parseInput(file string) ([]inputData, error) {
+// paserInput parses an input csv file to instances of inputData and puts them
```
```suggestion
// parseInput parses an input csv file to instances of inputData and puts them
```

ckyiricky replied:
fixed

robscott commented on process/process.go:
```diff
@@ -31,18 +30,24 @@ const inZoneTrafficScoreWeight, deviationScoreWeight, sliceScoreWeight = 0.4, 0.
 func StartProcessing(inputFile string, outputFile string, alg string) error {
-	inputArray, err := parseInput(inputFile)
+
+	// initialize a goroutine to keep reading row data from input file and
+	// putting the converted row data into a queue
```
```suggestion
	// initialize a goroutine to keep read row data from input file and
	// put the converted row data into a queue
```

ckyiricky replied:
I guess you meant 'goroutine to read xx and put' same as comments below. 
changed to 
'''
// initialize a goroutine to read row data from input file and
// put the converted row data into a queue
'''

robscott commented on process/process.go:
```diff
@@ -31,18 +30,24 @@ const inZoneTrafficScoreWeight, deviationScoreWeight, sliceScoreWeight = 0.4, 0.
 	}
 
-	outputArray, err := startSimulation(alg, inputArray)
+	// initialize a goroutine to keep processing row data from inputQueue and
+	// putting the processed data into another queue
```
```suggestion
	// initialize a goroutine to process row data from inputQueue and
	// put the processed data into another queue to handle results
```

ckyiricky replied:
changed

robscott commented on process/process.go:
```diff
@@ -31,18 +30,24 @@ const inZoneTrafficScoreWeight, deviationScoreWeight, sliceScoreWeight = 0.4, 0.
 
-	err = parseResult(outputFile, outputArray)
-	return err
+	// parse results from outputQueue and write to output file in the main
+	// thread which will end after all data has been processed
```
```suggestion
	// parse results from outputQueue and write to output file
```

ckyiricky replied:
changed

robscott commented on process/process.go:
```diff
@@ -66,33 +71,49 @@ type outputData struct {
-		if err != nil {
-			return outputArray, fmt.Errorf("error starting simulation for input : %s, %v", rowData.name, err)
+	outputQueue := make(chan outputData)
+	// since we only have one goroutine working on the data processing, although
+	// the model is not thread safe, for simplicity we create the model outside
+	// the goroutine but use it inside the goroutine.
```
```suggestion
	// Some simplifications here result in this code not being threadsafe.
	// Do not use more than one goroutine to process this queue.
```

ckyiricky replied:
changed

robscott reviewed (commented):
Mostly LGTM, just a few minor nits around wording.

ckyiricky commented:
Fix typos

robscott reviewed (approved):
Thanks for the work on this!
